src/test/scala/com/crealytics/spark/v2/excel/WriteAndReadSuite.scala

```diff
@@ -67,6 +72,30 @@ object WriteAndReadSuite {
     Row(1L, 12L, "CA872", "Phan Tố Nga", null, null, 2000, null, "Teresa Teng"),
     Row(1L, 12L, "CA873", "Nguyễn Thị Teresa Teng", null, null, 1200, null, "Jesse Thomas")
   ).asJava
+
+  val userDefinedSchema_03 = StructType(
+    List(
+      StructField("Id", IntegerType, nullable = true),
+      StructField("Date", DateType, nullable = true),
+      StructField("Timestamp", TimestampType, nullable = true)
+    )
+  )
+
+  val expectedData_03_sql: util.List[Row] = List(
+    Row(1, Date.valueOf("2021-10-01"), Timestamp.valueOf("2021-10-01 01:23:45")),
```

Comment:
You could store the data as `Tuple3[Int, String, String]` once and then convert the two `String` entries into the required datatypes in the corresponding tests. Also, no need to convert to a Java `util.List[Row]`. Spark has a [`.createDataFrame` version that takes a `Seq[A <: Product]`](https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/SparkSession.html#createDataFrame[A%3C:Product](data:Seq[A])(implicitevidence$3:reflect.runtime.universe.TypeTag[A]):org.apache.spark.sql.DataFrame) for which should work with a `List[Tuple3[...]]`.

Reply:
partially done. I wasn't able to get it working nicely with .createataFrame(Seq[A]) since that doesn't take a Struct and removing the Struct caused some other issues (after repplacing "Id" with "_1" as the column name)